drivers/mspi/mspi_dw.c

```diff
@@ -283,62 +529,93 @@ static bool read_rx_fifo(const struct device *dev,
 	const struct mspi_xfer_packet *packet =
 		&dev_data->xfer.packets[dev_data->packets_done];
 	bool finished = false;
 
-	if (packet->dir == MSPI_TX) {
-		if (dev_data->buf_pos < dev_data->buf_end) {
-			tx_data(dev, packet);
-		} else {
-			/* It may happen that at this point the controller is
-			 * still shifting out the last frame (the last interrupt
-			 * occurs when the TX FIFO is empty). Wait if it signals
-			 * that it is busy.
-			 */
-			while (read_sr(dev) & SR_BUSY_BIT) {
-			}
+	uint32_t int_status = read_isr(dev);
+#if defined(CONFIG_MSPI_DMA)
+	bool dma_irq = vendor_specific_read_dma_irq(dev);
 
-			finished = true;
-		}
+	if (xfer.xfer_mode == MSPI_DMA && dma_irq) {
+		/* No need to read FIFO by CPU */
+		finished = true;
 	} else {
-		uint32_t int_status = read_isr(dev);
-
-		do {
-			if (int_status & ISR_RXFIS_BIT) {
-				if (read_rx_fifo(dev, packet)) {
-					finished = true;
-					break;
+#else
+	{
+#endif
+		if (packet->dir == MSPI_TX) {
+			if (dev_data->buf_pos < dev_data->buf_end) {
+				if (int_status & ISR_TXEIS_BIT) {
+					tx_data(dev, packet);
 				}
-
-				if (read_risr(dev) & RISR_RXOIR_BIT) {
-					finished = true;
-					break;
+			} else {
+				/* It may happen that at this point the controller is
+				 * still shifting out the last frame (the last interrupt
+				 * occurs when the TX FIFO is empty). Wait if it signals
+				 * that it is busy.
+				 */
+				while (read_sr(dev) & SR_BUSY_BIT) {
 				}
 
-				int_status = read_isr(dev);
+				finished = true;
 			}
-
-			if (int_status & ISR_TXEIS_BIT) {
-				if (tx_dummy_bytes(dev)) {
-					/* All the required dummy bytes were
-					 * written to the FIFO; disable the TXE
-					 * interrupt, as it's no longer needed.
-					 */
-					write_imr(dev, IMR_RXFIM_BIT);
+		} else {
+			do {
+				if (int_status & ISR_RXFIS_BIT) {
+					if (read_rx_fifo(dev, packet)) {
+						finished = true;
+						break;
+					}
+
+					if (read_risr(dev) & RISR_RXOIR_BIT) {
+						finished = true;
+						break;
+					}
+
+					int_status = read_isr(dev);
 				}
 
-				int_status = read_isr(dev);
-			}
-		} while (int_status != 0);
-	}
+				if (int_status & ISR_TXEIS_BIT) {
+					if (tx_dummy_bytes(dev)) {
+						/* All the required dummy bytes were
+						 * written to the FIFO; disable the TXE
+						 * interrupt, as it's no longer needed.
+						 */
+						write_imr(dev, IMR_RXFIM_BIT);
+					}
 
+					int_status = read_isr(dev);
+				}
+			} while (int_status != 0);
+		}
+	}
 	if (finished) {
 		write_imr(dev, 0);
+		#if defined(CONFIG_MSPI_DMA)
+		/* Free DMA transfer list from previous packet if allocated */
+		if (dev_data->xfer.xfer_mode == MSPI_DMA && dev_data->dma_transfer_list) {
+			vendor_specific_free_dma_transfer_list(dev);
+		}
+		#endif
+		/* For async, minimal ISR - just submit work and return */
+		if (xfer.async) {
+			/* Cancel the timeout timer since transaction completed normally */
+			k_timer_stop(&dev_data->async_timeout_timer);
+			write_ssienr(dev, 0);
```

Comment:
Why is the controller disabled here? If XIP is enabled, this can lead to a crush.